File: l1/z7/recursive.c
#include <stdlib.h>
#include "recursive.h"

long factorial_recursive(int n) {
    PyObject *args = PyTuple_Pack(1, PyLong_FromLong(n));
    return call_python_function("recursive", "factorial_recursive", args);
}

long gcd_recursive(int a, int b) {
    PyObject *args = PyTuple_Pack(2, PyLong_FromLong(a), PyLong_FromLong(b));
    return call_python_function("recursive", "gcd_recursive", args);
}

ExtGCDResult solve_diophantine_recursive(int a, int b, int c) {
    return call_python_function_diophantine("recursive", "solve_diophantine_recursive", a, b, c);
}

File: l1/z7/recursive.py
from dataclasses import dataclass

@dataclass
class ExtGCDResult:
    gcd: int
    x: int
    y: int

def factorial_recursive(n):
    if n == 0:
        return 1
    else:
        return n * factorial_recursive(n-1)

def gcd_recursive(a, b):
    if b == 0:
        return a
    else:
        return gcd_recursive(b, a % b)

def extended_gcd_recursive(a, b):
    if b == 0:
        return ExtGCDResult(gcd=a, x=1, y=0)
    else:
        temp = extended_gcd_recursive(b, a % b)
        return ExtGCDResult(gcd=temp.gcd, x=temp.y, y=temp.x - (a // b) * temp.y)

def solve_diophantine_recursive(a, b, c):
    result = extended_gcd_recursive(a, b)
    if c % result.gcd != 0:
        return "No solution"
    result.x *= c // result.gcd
    result.y *= c // result.gcd
    return ExtGCDResult(gcd=result.gcd, x=result.x, y=result.y)